### PR TITLE
feat: publish OpenTelemetry support to NuGet listed latest version.

### PR DESCRIPTION
**Pull request by guitarrapc**
closed https://github.com/Cysharp/MagicOnion/issues/421

## TL;DR;

* CI: Enable CI and Release for Experimental package.
* BreakingChange: Drop Metrics support as OpenTelemetry-dotnet 1.0.1 and higher does. (will be re-support in .NET 6,) 
* BreakingChange: Split `IMagicOnionServerBuilder.AddOpenTelemetry()` to Listener `IMagicOnionServerBuilder.AddOpenTelemetry()` and Instrumentation `TracerProviderBuilder.AddMagicOnionInstrumentation()`. This offer Standard OpenTelemetry usage and simpler MagicOnion Integration. (see sample code)
* Enahance: Support OpenTelemetry Tracer Span Context Propagation for Server to Server scenario.
* Package: Update MagicOnion.OpenTelemetry package suffix to `beta2-1.1.0` which represent `OpenTelemery 1.1.0-beta2`.
Nov/2021)
* Package: Drop OpenTelemetry.Extensions.Hosting reference, No need wait package update anymore.

## Summary

**BreakingChange: Drop Metrics support**

OpenTelemetry 1.0.1 is released for Tracer packages, however Metrics are dropped until .NET 6 release.
Follow to OpenTelemetry package, MagicOnion drop Metrics support.

**BreakingChange: Split Listener and Instrumentations**

Previously MagicOnion.OpenTelemetry integrate "Listen Activity" and "Enable Instrumentation" in single `IMagicOnionServerBuilder.AddOpenTelemetry()`. However this prevent user from standard OpenTelemetry usage.
Now it's separate to activity listener and enable instrumentation.

* `IMagicOnionServerBuilder.AddOpenTelemetry()` listen Activity and DI MagicOnionTelemetryOptions.
* `TracerProviderBuilder.AddMagicOnionInstrumentation()` enable MagicOnion Instrumentations.

before

```cs
 services.AddMagicOnion(options =>
      {
          options.GlobalFilters.Add(new MagicOnionOpenTelemetryTracerFilterFactoryAttribute());
          options.GlobalStreamingHubFilters.Add(new MagicOnionOpenTelemetryStreamingTracerFilterFactoryAttribute());

          // Exception Filter is inside telemetry
          options.GlobalFilters.Add(new ExceptionFilterFactoryAttribute());
      })
      .AddOpenTelemetry((options, provider, tracerBuilder) =>
      {
          tracerBuilder
              .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MyService"))
              .AddAspNetCoreInstrumentation()
              .AddJaegerExporter();
          services.Configure<OpenTelemetry.Exporter.JaegerExporterOptions>(Configuration.GetSection("Jaeger"));
      });
```

after
```cs
// Register MagicOnion Filter for OpenTelemetry && Listen OpenTelemetry Activity
services.AddMagicOnion(options =>
    {
        options.GlobalFilters.Add(new MagicOnionOpenTelemetryTracerFilterFactoryAttribute());
        options.GlobalStreamingHubFilters.Add(new MagicOnionOpenTelemetryStreamingTracerFilterFactoryAttribute());

        options.GlobalFilters.Add(new ExceptionFilterFactoryAttribute());
    })
    .AddOpenTelemetry(); // Listen Activity

// choice1. Enable MagicOnion instrumentation via IServiceCollection.AddOpenTelemetryTracing integration.
services.AddOpenTelemetryTracing(configure => 
{ 
    configure
        .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MyService"))
        .AddAspNetCoreInstrumentation()
        .AddMagicOnionInstrumentation() // enable MagicOnion instrumentation
        .AddJaegerExporter();
    services.Configure<OpenTelemetry.Exporter.JaegerExporterOptions>(Configuration.GetSection("Jaeger"));
});

// choice2. Enable MagicOnion instrumentation via TracerProviderBuilder Sdk
var jo = new OpenTelemetry.Exporter.JaegerExporterOptions();
configuration.GetSection("Jaeger").Bind(jo);
Sdk.CreateTracerProviderBuilder()
    .AddAspNetCoreInstrumentation()
    .AddMagicOnionInstrumentation() // enable MagicOnion instrumentation
    .AddJaegerExporter(o =>
    {
        o.AgentHost = jo.AgentHost;
        o.AgentPort = jo.AgentPort;
    })
    .Build();
```

**Enahance: Support OpenTelemetry Context Propagation**

Now OpenTelemetry Context will propagate from server to server. This is done via gRPC Request Header.
This enable external MagicOnion Service Span integrate to single TraceId and under caller Spanid.
Example shows external MagicOnion `MicroServer` is called from `ChatApp.Server`'s `ChatApp.Server.S2S` operation.

image

```csharp
// Caller ChatApp.Server Propagate Context to callee.
public class ChatService : ServiceBase<IChatService>, IChatService
{
    public async UnaryResult<Nil> SendReportAsync(string message)
    {
        logger.LogDebug($"{message}");

        // dummy external operation.

        // Server to Server operation
        var channel = GrpcChannel.ForAddress(Environment.GetEnvironmentVariable("Server2ServerEndpoint", EnvironmentVariableTarget.Process) ??  "http://localhost:4999");
        var client = MagicOnionClient.Create<IMessageService>(channel, new[]
        {
            // propagate trace context from ChatApp.Server to MicroServer
            new MagicOnionOpenTelemetryClientFilter(s2sActivity, options),
        });
        await client.SendAsync("hello");

        // dummy external operation.
    }
}

// callee MicroServer automatically obtain propagated OpenTelemetry Span Context through Request Header.
public class MessageService : ServiceBase<IMessageService>, IMessageService
{
    public async UnaryResult<string> SendAsync(string message)
    {
        await Task.Delay(TimeSpan.FromMilliseconds(20));
        return $"echo {message}";
    }
}
```

## REF

> [Remove metrics from 1\.0\.0 release by cijothomas · Pull Request \#1743 · open\-telemetry/opentelemetry\-dotnet](https://github.com/open-telemetry/opentelemetry-dotnet/pull/1743)
> [Metric support plans · Issue \#1501 · open\-telemetry/opentelemetry\-dotnet](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1501)

### CODE REVIEW

**Comment by guitarrapc** [comment]
Linux has Tracer not sending issue.

> https://github.com/open-telemetry/opentelemetry-dotnet/issues/2050

**Comment by guitarrapc** [comment]
downgrade OpenTElemetry 1.1.0-beta2 can send trace on linux, but exporters install with `Sdk.CreateTracerProviderBuilder()` like mysql and others will not.

**Comment by guitarrapc** [comment]
All know issues are solved by Splitting Instrumentation.

**Review by mayuki — APPROVED** [review]
